google/cloud/storage/status_or.h

```diff
@@ -204,6 +204,117 @@ class StatusOr final {
   T value_;
 };
 
+/**
+ * `StatusOr<T>` is used to return a value or an error status when exceptions
+ * are disabled.
+ *
+ * If the library is compiled with exceptions disabled there is no standard C++
+ * mechanism to report errors. In that case we use this class to wrap the values
+ * returned to the application.
+ *
+ * The application typically calls an API in the library and must check the
+ * returned error status:
+ *
+ * @code
+ * namespace gcs = google::cloud::storage;
+ * void AppCode(gcs::noex::Client client) {
+ *   gcs::StatusOr<gcs::BucketMetadata> meta_err = client.GetBucketMetadata(
+ *       "my-bucket-name");
+ *   if (not meta_err.ok()) {
+ *       std::cerr << "Error in GetBucketMetadata: " << meta_err.status()
+ *                 << std::endl;
+ *       return;
+ *   }
+ *   gcs::BucketMetadata meta = std::move(meta_err).value();
+ *   // Do useful work here.
+ * }
+ * @endcode
+ *
+ * Note that the storage client retries most requests for you, resending the
+ * request after an error is probably not useful. You should consider changing
+ * the retry policies instead.
+ *
+ * TODO(...) - the current implementation is fairly naive with respect to `T`,
+ *   it is unlikely to work correctly for reference types, types without default
+ *   constructors, arrays.
+ *
+ * @tparam T the type of the value.
```

Comment:
These comments are mostly duplicated from the StatusOr primary template above. What do you think about instead having these comments only discuss what's "special" about this specialization? That is, it has no constructor from `void`, etc.?

Reply:
Done.